Hummingbird.docc/Articles/Recommendations.md

```diff
@@ -0,0 +1,342 @@
+- Job Queues
+- Caching
+- OpenTelemetry
+- File System Access
+
+Each service manages it's lifecycle through [swift-service-lifecycle](https://github.com/swift-server/swift-service-lifecycle). The order of services is important, as services are initialized in the order they are added to the application, and teared down in reverse order.
+
+```swift
+let app = Application()
+let postgres = PostgresService(connectionString: arguments.postgres)
+app.addServices(
+    OpenTelemetryService(),
+    // Postgres reports to OTel through the global logger, so initialized later
+    postgres,
+    // Depends on postgres, so initialized later
+    PostgresJobQueueService(postgres: postgres)
+)
+```
+
+## Observability
+
+Swift's Observability APIs offer a powerful and flexible way to instrument and observe the application. Hummingbird already reports to [swift-log](https://github.com/apple/swift-log) and provides middleware for [swift-metrics](https://github.com/apple/swift-metrics) and [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing). These will report to the globally set up observability backend(s) automatically, if they're set up.
+
+By default, logs are emitted to `stdout`, and metrics + traces are discarded. If you want to trace signals in a single place, we recommend using [swift-otel](https://github.com/swift-otel/swift-otel/issues) as an observability backend.
+
+See the [OTel Example Project](https://github.com/hummingbird-project/hummingbird-examples/tree/main/open-telemetry) for a complete example.
+
+### Logging Recommendations
+
+Use the [Logging Guidelines](https://www.swift.org/documentation/server/guides/libraries/log-levels.html) set forth by the Swift Server Workgroup. It's strongly recommended to leverage **structured logging** to make logs more readable and searchable.
+
+✅ Good:
+
+```swift
+logger.info("User logged in", metadata: ["user_id": .string(user.id)])
+```
+
+❌ Bad:
+
+```swift
+logger.info("User \(user.id) logged in")
+```
+
+## Scalability
+
+Hummingbird is explicitly designed for resource efficiency. A good example is how HTTP bodies are always represented as an `AsyncSequence` of bytes. This means that the server can stream data to the client without having to buffer the entire response in memory. Likewise, any uploads from a user through Hummingbird can be efficiently handled through the `AsyncSequence` APIs.
+
+It's highly recommended that you avoid `collect`ing responses into memory if possible. Instead, attempt to leverage the `AsyncSequence` APIs to stream data.
+
+Any data that needs to be stored in contiguous memory, like a JSON blob, should be strictly limited in the maximum size as to avoid (memory) resource exhaustion and performance issues.
+
+### Handling Bodies
+
+If you're using standard libraries like Foundation's `JSONDecoder`, limit the body size to "reasonable" sizes. What is considered reasonable depends per application, but lower is generally better.
```

Comment:
A lot of this is covered in the router documentation.
Recommendations about not collecting response in middleware, setting realistic limits when collecting into a single buffer, although preferring to stick to streamed version if possible.

Reply:
I'll make things a little more bullet-pointy here